BUG: rolling.count with axis=1

yhaque1213 opened the pull request:
- [ ] closes #13503 
- [X] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry
Added axis parameter to the constructor call in the count function, fixing the original issue in #13503. This fix works for the example given in the issue. This does not fix the problem addressed by jorisvandenbossche, as that seems to be a type issue.

yhaque1213 committed:
Added axis to constructor call in count

yhaque1213 committed:
Added axis to constructor call in count

yhaque1213 committed:
Added axis to constructor call in count

jreback commented:
tests pls

yhaque1213 committed:
TST: GH26055 added test_count_axis to TestRolling

pep8speaks commented:
Hello @yhaque1213! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/tests/test_window.py`](https://github.com/pandas-dev/pandas/blob/c64c76a30d1bd32b415010a297b5af6e15332445/pandas/tests/test_window.py):

> [Line 672:31](https://github.com/pandas-dev/pandas/blob/c64c76a30d1bd32b415010a297b5af6e15332445/pandas/tests/test_window.py#L672): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 672:45](https://github.com/pandas-dev/pandas/blob/c64c76a30d1bd32b415010a297b5af6e15332445/pandas/tests/test_window.py#L672): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 674:49](https://github.com/pandas-dev/pandas/blob/c64c76a30d1bd32b415010a297b5af6e15332445/pandas/tests/test_window.py#L674): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 674:80](https://github.com/pandas-dev/pandas/blob/c64c76a30d1bd32b415010a297b5af6e15332445/pandas/tests/test_window.py#L674): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (108 > 79 characters)
> [Line 679:52](https://github.com/pandas-dev/pandas/blob/c64c76a30d1bd32b415010a297b5af6e15332445/pandas/tests/test_window.py#L679): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 679:80](https://github.com/pandas-dev/pandas/blob/c64c76a30d1bd32b415010a297b5af6e15332445/pandas/tests/test_window.py#L679): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (111 > 79 characters)
> [Line 680:80](https://github.com/pandas-dev/pandas/blob/c64c76a30d1bd32b415010a297b5af6e15332445/pandas/tests/test_window.py#L680): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)

gfyoung commented on pandas/tests/test_window.py:
```diff
@@ -667,6 +667,18 @@ def test_rolling_axis(self, axis_frame):
+        assert_equal(df.rolling(2, axis='rows').count(), df_correct_row_count)
+
+        df_correct_column_count = pd.DataFrame({'x':[1.0, 1.0, 1.0, 1.0, 1.0], 'y': [2.0, 2.0, 2.0, 2.0, 2.0]})
+        assert_equal(df.rolling(2, axis='columns').count(), df_correct_column_count)
```
Use `tm.assert_frame_equal` instead of `assert_equal`.

yhaque1213 committed:
TST: fixed asserts and pep8 for test_count_axis()

yhaque1213 commented:
@gfyoung 
I updated the tests to replace my calls to `assert_equal` with `tm.assert_frame_equal`, and I also modified it to pass pep8 checks. My initial changes to fix the issue, without any tests, was passing all of the builds.

Once I added my tests, however, I started failing pandas-dev.pandas and pandas-dev.pandas (Linux py35_compat), and I'm having some trouble figuring out what's causing this to happen.

More specifically, it seems to be an error with the `test_drop_duplicates_categorical_non_bool` in `pandas/tests/series/test_analytics.py`, but I have no idea why this specific test would be failing because of adding my test, `test_count_axis`. 

Do you have any advice?

gfyoung commented:
@yhaque1213 : Have you rebased onto the most recent `master`?  Try that and see if it goes away.  Otherwise, we'll look into seeing if we can remove (or weaken) the `xfail` on those tests.

yhaque1213 committed:
TST: GH26055 rebasing with master

yhaque1213 commented:
@gfyoung : that seemed to do the trick - my latest update is passing all of the checks. Thank you, and please let me know if there's anything else you'd like me to add!

gfyoung commented:
@yhaque1213 : Don't forget the `whatsnew` entry

jreback commented on pandas/tests/test_window.py:
```diff
@@ -667,6 +667,18 @@ def test_rolling_axis(self, axis_frame):
+        df_row = DataFrame({'x': [1.0, 2.0, 2.0], 'y': [1.0, 2.0, 2.0]})
+
+        # not specifiying axis and making axis=rows should be the same result
+        tm.assert_frame_equal(df.rolling(2).count(), df_row)
```
use 

result=
expected=
tm.assert_frame_equal

so df_row looks like your expected

jreback commented on pandas/tests/test_window.py:
```diff
@@ -667,6 +667,18 @@ def test_rolling_axis(self, axis_frame):
+        tm.assert_frame_equal(df.rolling(2).count(), df_row)
+        tm.assert_frame_equal(df.rolling(2, axis='rows').count(), df_row)
+
+        df_col = DataFrame({'x': [1.0, 1.0, 1.0], 'y': [2.0, 2.0, 2.0]})
```
same here

jreback commented on pandas/tests/test_window.py:
```diff
@@ -667,6 +667,18 @@ def test_rolling_axis(self, axis_frame):
         result = df.rolling(3, axis=axis_frame).sum()
         tm.assert_frame_equal(result, expected)
 
+    def test_count_axis(self):
```
rename this to test_rolling_axis_count, rename the test right above to test_rolling_axis_sum

jreback reviewed (changes requested):
pls add a whatsnew note as well; bug fixes groupby/rolling section

yhaque1213 committed:
TST: GH26055 added whatsnew entry and renamed result and expected variables

yhaque1213 committed:
TST: GH26055 corrected doc errors and test pep8 errors

yhaque1213 committed:
TST: GH26055 fixed comment typo

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -384,6 +384,7 @@ Groupby/Resample/Rolling
 - Ensured that result group order is correct when grouping on an ordered ``Categorical`` and specifying ``observed=True`` (:issue:`25871`, :issue:`25167`)
 - Bug in :meth:`pandas.core.window.Rolling.min` and :meth:`pandas.core.window.Rolling.max` that caused a memory leak (:issue:`25893`)
 - Bug in :func:`idxmax` and :func:`idxmin` on :meth:`DataFrame.groupby` with datetime column would return incorrect dtype (:issue:`25444`, :issue:`15306`)
+- Bug in :meth:`pandas.core.window._Rolling_and_Expanding.count` that did not include axis parameter in constructor call (:issue:`13503`)
```
this is not a public method, use ``:meth:`pandas.core.window.Rolling.count`` and another for Expanding

jreback replied:
this is also confusing over what you are saying, can you reword, something like '....was previously ignoring the ``axis`` keyword'

jreback commented on pandas/tests/test_window.py:
```diff
@@ -667,6 +667,23 @@ def test_rolling_axis(self, axis_frame):
         result = df.rolling(3, axis=axis_frame).sum()
         tm.assert_frame_equal(result, expected)
 
+    def test_rolling_axis_count(self):
```
use the axis_frame fixture which resoles to rows/columns/0/1

yhaque1213 committed:
TST: GH26055 fixed docs description and added axis_frame to test

yhaque1213 commented:
I made the changes to the test and to the docs, but it seems like these Azure pipelines checks are passing arbitrarily. For instance, after fixing a typo in a comment (commit #12065b6) we passed 10/10 tests, while the previous commit only passed 8 (commit #41ae7fe). This happened even with rebasing before each push, therefore, we aren't sure what to change for this last commit to get it to pass all 10.

jreback commented:
> I made the changes to the test and to the docs, but it seems like these Azure pipelines checks are passing arbitrarily. For instance, after fixing a typo in a comment (commit #12065b6) we passed 10/10 tests, while the previous commit only passed 8 (commit #41ae7fe). This happened even with rebasing before each push, therefore, we aren't sure what to change for this last commit to get it to pass all 10.

CI is not 'arbitrary', rather there was an upstream change that broke things.

jreback commented:
pls merge master and ping on green.

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -384,6 +384,8 @@ Groupby/Resample/Rolling
 - Bug in :meth:`pandas.core.window.Rolling.min` and :meth:`pandas.core.window.Rolling.max` that caused a memory leak (:issue:`25893`)
 - Bug in :func:`idxmax` and :func:`idxmin` on :meth:`DataFrame.groupby` with datetime column would return incorrect dtype (:issue:`25444`, :issue:`15306`)
+- Bug in :meth:`pandas.core.window.Rolling.count` was previously ignoring the axis keyword (:issue:`13503`)
+- Bug in :meth:`pandas.core.window.Expanding.count` was previously ignoring the axis keyword (:issue:`13503`)
```
make a single entry here (just combine the lines)

jreback commented on pandas/tests/test_window.py:
```diff
@@ -667,6 +667,21 @@ def test_rolling_axis(self, axis_frame):
+
+        axis = df._get_axis_number(axis_frame)
+
+        if axis == 0:
```
this is not correct (should your test have broken?)

``if axis in [0, 'index']`` is what you need

yhaque1213 committed:
Combined description in doc and modified if statement in test to correct the axis check

yhaque1213 committed:
TST: GH26055 Same changes as previous commit, merge did not pass checks, trying rebase

yhaque1213 committed:
Merge branch 'master' into count-axis-fix

yhaque1213 commented:
@jreback : Just made the changes requested, and I'm passing all the checks now!

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -390,6 +390,8 @@ Groupby/Resample/Rolling
 - Ensured that ordering of outputs in ``groupby`` aggregation functions is consistent across all versions of Python (:issue:`25692`)
 - Ensured that result group order is correct when grouping on an ordered ``Categorical`` and specifying ``observed=True`` (:issue:`25871`, :issue:`25167`)
 - Bug in :meth:`pandas.core.window.Rolling.min` and :meth:`pandas.core.window.Rolling.max` that caused a memory leak (:issue:`25893`)
+- Bug in :func:`idxmax` and :func:`idxmin` on :meth:`DataFrame.groupby` with datetime column would return incorrect dtype (:issue:`25444`, :issue:`15306`)
```
looks like this was picked up somehow, pls remove

jreback reviewed (changes requested):
looks good; edit the whatsnew; ping on green.

yhaque1213 committed:
CLN: GH26055 fixed doc and undid unintentionally added commits

yhaque1213 committed:
DOC: GH26055 removed extra doc line

yhaque1213 committed:
DOC: GH26055 removed extra doc lines

yhaque1213 committed:
DOC: GH26055 removed extra doc lines

yhaque1213 committed:
DOC: GH26055 readded accidentally removed doc lines

yhaque1213 committed:
DOC: GH26055 readded accidentally removed doc lines

yhaque1213 committed:
DOC: GH26055 removed extra space

yhaque1213 committed:
DOC: GH26055 removed extra space

yhaque1213 committed:
DOC: GH26055 merged doc conflicts

yhaque1213 committed:
TST: GH26055 merging test conflicts

yhaque1213 committed:
BUG: GH26055 removed extra assignment in constructor call

yhaque1213 committed:
BUG: GH26055 removed extra assignment in constructor call

yhaque1213 committed:
DOC: GH26055 removed rebase message

yhaque1213 committed:
DOC: GH26055 removed rebase message

yhaque1213 commented:
I've made all of the changes that were requested, but I'm still having trouble getting all of the checks to pass. I've tried rebasing and merging multiple times, but nothing has been working

jreback commented:
try

git merge upstream/master

WillAyd commented:
There's a failure in the doc checks that should be unrelated. Not sure if its a versioning thing or a commit to master. Taking a look now

jreback commented:
this was cleared yesterday

yhaque1213 committed:
DOC: GH26055 merging doc conflicts

yhaque1213 committed:
DOC: GH26055 updated docs to reserve merge conflict

yhaque1213 commented:
Latest push was with merging with upstream/master, but it looks like the checks are still failing in the same ways as before. Sorry I'm not sure if I'm doing something wrong, but I've tried everything I could think of

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -387,6 +387,7 @@ Groupby/Resample/Rolling
 - Ensured that ordering of outputs in ``groupby`` aggregation functions is consistent across all versions of Python (:issue:`25692`)
 - Ensured that result group order is correct when grouping on an ordered ``Categorical`` and specifying ``observed=True`` (:issue:`25871`, :issue:`25167`)
 - Bug in :meth:`pandas.core.window.Rolling.min` and :meth:`pandas.core.window.Rolling.max` that caused a memory leak (:issue:`25893`)
+- Bug in :meth:`pandas.core.window.Rolling.count` and `pandas.core. window.Expanding.count` was previously ignoring the axis keyword (:issue:`13503`)
```
can you remove the space here pandas.core.window....

jreback reviewed (changes requested):
lgtm. tiny typo. pls rebase on master which is green now.

jreback commented:
small typo plus can i merge master ; ping on green

yhaque1213 committed:
DOC: GH26055 removed typo from doc

yhaque1213 committed:
DOC: GH26055 removed typo

yhaque1213 commented:
@jrebeck: I removed the typo - I'm passing all the checks now!

jreback commented:
thanks @yhaque1213
